File: app/src/main/java/ru/deniskrd/android/webbrowser/PropertiesHelper.java
package ru.deniskrd.android.webbrowser;

import android.content.Context;

import java.util.Properties;

import ru.deniskrd.android.webbrowser.db.AppDatabase;
import ru.deniskrd.android.webbrowser.db.entities.Favourite;
import ru.deniskrd.android.webbrowser.db.entities.UserSetting;

class PropertiesHelper {

    private Properties userSettings;
    private Properties favourites;

    private AppDatabase appDatabase;

    PropertiesHelper(Context context) {
        appDatabase = AppDatabase.getInstance(context);
    }

    Properties getUserSettings() {
        return userSettings;
    }

    Properties getFavourites() {
        return favourites;
    }

    void loadProperties() {
        userSettings = new Properties();
        favourites = new Properties();

        appDatabase.getQueryExecutor().execute(new Runnable() {
            @Override
            public void run() {
                for (UserSetting userSetting : appDatabase.userSettingDao().getAllUserSettings()) {
                    userSettings.setProperty(userSetting.getName(), userSetting.getValue());
                }

                for (Favourite favourite : appDatabase.favouriteDao().getAllFavourites()) {
                    favourites.setProperty(favourite.getName(), favourite.getUrl());
                }
            }
        });
    }

    void saveProperties() {
        appDatabase.getTransactionExecutor().execute(new Runnable() {
            @Override
            public void run() {
                appDatabase.clearAllTables();

                for (String settingName : userSettings.stringPropertyNames()) {
                    appDatabase.userSettingDao().setUserSetting(new UserSetting(settingName, userSettings.getProperty(settingName)));
                }

                for (String favouriteName : favourites.stringPropertyNames()) {
                    appDatabase.favouriteDao().setFavourite(new Favourite(favouriteName, favourites.getProperty(favouriteName)));
                }
            }
        });
    }
}

File: app/src/main/java/ru/deniskrd/android/webbrowser/PropertiesHelper.kt
package ru.deniskrd.android.webbrowser;

import android.content.Context
import io.reactivex.Observable
import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import ru.deniskrd.android.webbrowser.db.AppDatabase
import ru.deniskrd.android.webbrowser.db.entities.Favourite
import ru.deniskrd.android.webbrowser.db.entities.UserSetting
import java.util.*

class PropertiesHelper() {

    val userSettings = Properties()

    val favourites = Properties()

    private var appDatabase: AppDatabase? = null

    constructor(context: Context) : this() {
        appDatabase = AppDatabase.getInstance(context)
    }

    fun loadProperties() {
        appDatabase!!.queryExecutor.execute {
            appDatabase!!.userSettingDao().getAllUserSettings().forEach { userSetting ->
                userSettings.setProperty(userSetting.name, userSetting.value)
            }
            appDatabase!!.favouriteDao().getAllFavourites().forEach { favourite ->
                favourites.setProperty(favourite.name, favourite.url)
            }
        }
    }

    fun saveProperties() {
        appDatabase!!.transactionExecutor.execute {
            appDatabase!!.clearAllTables()

            userSettings.forEach { (name, value) ->
                appDatabase!!.userSettingDao().setUserSetting(UserSetting(name as String, value as String))
            }

            favourites.forEach { (name, value) ->
                appDatabase!!.favouriteDao().setFavourite(Favourite(name as String, value as String))
            }
        }
    }
}
